daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/SubmitError.scala

```diff
@@ -0,0 +1,295 @@
+  ) extends SubmitError {
+    override def toDamlSubmitError(env: Env): SValue =
+      SubmitErrorConverters(env).damlScriptError(
+        "ContractDoesNotImplementInterface",
+        14,
+        ("contractId", fromAnyContractId(env.scriptIds, toApiIdentifier(templateId), contractId)),
+        ("templateId", fromTemplateTypeRep(toApiIdentifier(templateId))),
+        ("requiredInterfaceId", fromTemplateTypeRep(toApiIdentifier(requiredInterfaceId))),
+        ("requiringInterfaceId", fromTemplateTypeRep(toApiIdentifier(requiringInterfaceId))),
+      )
+  }
+
+  final case class NonComparableValues() extends SubmitError {
+    override def toDamlSubmitError(env: Env): SValue =
+      SubmitErrorConverters(env).damlScriptError(
+        "NonComparableValues",
+        15,
+      )
+  }
+
+  final case class ContractIdInContractKey() extends SubmitError {
+    override def toDamlSubmitError(env: Env): SValue =
+      SubmitErrorConverters(env).damlScriptError(
+        "ContractIdInContractKey",
+        16,
+      )
+  }
+
+  final case class ContractIdComparability(contractId: String) extends SubmitError {
+    override def toDamlSubmitError(env: Env): SValue =
+      SubmitErrorConverters(env).damlScriptError(
+        "ContractIdComparability",
+        17,
+        ("globalExistingContractId", SText(contractId)),
+      )
+  }
+
+  final case class DevError(errorType: String, message: String) extends SubmitError {
+    override def toDamlSubmitError(env: Env): SValue = {
+      val devErrorTypeIdentifier =
+        env.scriptIds.damlScriptModule("Daml.Script.Questions.Submit.Error", "DevErrorType")
+      val devErrorType = errorType match {
+        case "ChoiceGuardFailed" =>
+          SEnum(devErrorTypeIdentifier, Name.assertFromString("ChoiceGuardFailed"), 0)
+        case "WronglyTypedContractSoft" =>
+          SEnum(devErrorTypeIdentifier, Name.assertFromString("WronglyTypedContractSoft"), 1)
+        case "Limit" => SEnum(devErrorTypeIdentifier, Name.assertFromString("Limit"), 2)
+        case _ => SEnum(devErrorTypeIdentifier, Name.assertFromString("UnknownNewFeature"), 3)
+      }
+      SubmitErrorConverters(env).damlScriptError(
+        "DevError",
+        18,
+        ("devErrorType", devErrorType),
+        ("devErrorMessage", SText(message)),
+      )
+    }
+  }
+
+  final case class UnknownError(message: String) extends SubmitError {
+    override def toDamlSubmitError(env: Env): SValue =
+      SubmitErrorConverters(env).damlScriptError(
+        "UnknownError",
+        19,
+        ("unknownErrorMessage", SText(message)),
+      )
+  }
+
+  final case class TruncatedError(errType: String, message: String) extends SubmitError {
+    override def toDamlSubmitError(env: Env): SValue =
+      SubmitErrorConverters(env).damlScriptError(
+        "TruncatedError",
+        20,
+        ("truncatedErrorType", SText(errType)),
+        ("truncatedErrorMessage", SText(message)),
+      )
+  }
+}
```

Comment:
Phew that's a big plate of boiler, thanks for getting this one through.

Reply:
Yeah - i wish it could be refactored to less, but I'm not convinced it can